community/lms/doctype/exercise/exercise.py

```diff
@@ -55,5 +56,9 @@ def submit(self, code):
             image=image,
             solution=code)
         doc.insert(ignore_permissions=True)
+
+        if not (course.is_mentor(frappe.session.user) or frappe.flags.in_test):
+            update_progress(self.lesson)
```

Comment:
I guess you need to call `update_progress` on `ExerciseSubmission.before_save` and not  here.